tfhe/src/integer/gpu/list_compression/server_keys.rs

```diff
@@ -0,0 +1,239 @@
+            storage_log_modulus: compression_key.storage_log_modulus,
+        }
+    }
+
+    unsafe fn flatten_async(
+        vec_ciphertexts: &Vec<CudaRadixCiphertext>,
+        streams: &CudaStreams,
+    ) -> (CudaLweCiphertextList<u64>, Vec<CudaBlockInfo>) {
+        let first_ct = &vec_ciphertexts.first().unwrap().d_blocks;
+
+        // We assume all ciphertexts will have the same lwe dimension
+        let lwe_dimension = first_ct.lwe_dimension();
+        let ciphertext_modulus = first_ct.ciphertext_modulus();
+
+        // Compute total number of lwe ciphertexts we will be handling
+        let total_num_blocks: usize = vec_ciphertexts
+            .iter()
+            .map(|x| x.d_blocks.lwe_ciphertext_count().0)
+            .sum();
+
+        let lwe_ciphertext_count = LweCiphertextCount(total_num_blocks);
+
+        let gpu_index = streams.gpu_indexes[0];
+        let d_vec = unsafe {
+            let mut d_vec = CudaVec::new_async(
+                lwe_dimension.to_lwe_size().0 * lwe_ciphertext_count.0,
+                streams,
+                gpu_index,
+            );
+            let mut offset: usize = 0;
+            for ciphertext in vec_ciphertexts {
+                // Todo: We might use copy_self_range_gpu_to_gpu_async here
+                let dest_ptr = d_vec
+                    .as_mut_c_ptr(gpu_index)
+                    .add(offset * std::mem::size_of::<u64>());
+                let size = ciphertext.d_blocks.0.d_vec.len * std::mem::size_of::<u64>();
+                cuda_memcpy_async_gpu_to_gpu(
+                    dest_ptr,
+                    ciphertext.d_blocks.0.d_vec.as_c_ptr(gpu_index),
+                    size as u64,
+                    streams.ptr[gpu_index as usize],
+                    streams.gpu_indexes[gpu_index as usize],
+                );
+
+                offset += ciphertext.d_blocks.0.d_vec.len;
+            }
+
+            streams.synchronize();
+            d_vec
+        };
+
+        let flattened_ciphertexts =
+            CudaLweCiphertextList::from_cuda_vec(d_vec, lwe_ciphertext_count, ciphertext_modulus);
+
+        let info = vec_ciphertexts
+            .iter()
+            .flat_map(|x| x.info.blocks.clone())
+            .collect_vec();
+
+        (flattened_ciphertexts, info)
+    }
+
+    pub fn compress_ciphertexts_into_list(
+        &self,
+        ciphertexts: &Vec<CudaRadixCiphertext>,
+        streams: &CudaStreams,
+    ) -> (
+        CudaGlweCiphertextList<u64>,
+        Vec<CudaBlockInfo>,
+        CiphertextModulusLog,
+    ) {
+        let lwe_pksk = &self.packing_key_switching_key;
+
+        let ciphertext_modulus = lwe_pksk.ciphertext_modulus();
+        let compress_polynomial_size = lwe_pksk.output_polynomial_size();
+        let compress_glwe_size = lwe_pksk.output_glwe_size();
+
+        let first_ct = ciphertexts.first().unwrap();
+        let first_ct_info = first_ct.info.blocks.first().unwrap();
```

Comment:
what data is contained in the info ?

does it need a check here to verify all the infos are coherent between the ciphertexts ?

Reply:
This struct is contained in the info vec:

```
#[derive(Clone, Copy)]
pub struct CudaBlockInfo {
    pub degree: Degree,
    pub message_modulus: MessageModulus,
    pub carry_modulus: CarryModulus,
    pub pbs_order: PBSOrder,
    pub noise_level: NoiseLevel,
}
```

Maybe. Maybe not. The CPU does something similar but with an array of ciphertexts. It takes the first one and assume all the others will have the same values.

In both cases if we don't have coherency computation will fail. I would say any check would only be needed during the ciphertext creation and not here.

Reply:
ok I think the requirement is that the ciphertexts are fresh, to be confirmed by @mayeul-zama when he is back, as this must have been the atomic pattern used for compression

Reply:
Indeed, we expect ciphertexts to be clean (empty carries and fresh noise) before compression